### PR TITLE
Connect to docker with no timeout on startup

### PR DESCRIPTION
**Pull request by jsravn**
For https://github.com/google/cadvisor/issues/1866.

### CODE REVIEW

**Comment by dashpole** [comment]
cadvisor tests are not working right now.  We are in the process of migrating them off of jenkins.

**Comment by tn-osimis** [comment]
I could be wrong but this might mean that if the engine never replies for some reason cadvisor will hang indefinitely without giving a clue as to what it's trying to do in the logs.

It's a very welcome fix for #1866 and should probably be merged; though as a long-term approach I suggest a higher default timeout at startup (e.g. 30s or 60s) as well as the addition of a setting for configuring a non-default value.

(Otherwise, maybe at least add log output *before* attempting to connect so that ops can more easily figure out what's blocking the startup process in this scenario.)

WDYT?

**Comment by jsravn** [comment]
@tn-osimis yes, I agree. It is a more systemic problem in cadvisor though, it doesn't use timeouts for anything. So it should really be fixed throughout. I'd prefer using a larger timeout than indefinite timeout on startup.

**Comment by dashpole** [comment]
@jsravn I agree with @tn-osimis that just using a longer timeout would be preferable.  What I meant in the issue by "infinite retry", is actually using a call with a long timeout (e.g. 30 sec), but retrying the call if the error is a timeout.  This way, as soon as docker recovers, cAdvisor starts correctly.  If docker is permanently hung, we have bigger problems than missing metrics.  WDYT?

**Comment by tn-osimis** [comment]
If there is a generic library implementing an exponential backoff algorithm (or whatever else) somewhere in your stack already, that's probably what you want (and the logical conclusion to your thought). Otherwise indeed a basic fixed-interval retry approach usually works better than just waiting indefinitely for cases where the state machine of the peer (i.e. Docker engine) is a little messed up.

FWIW (opinion of a random guy on the Internet) I agree with retrying indefinitely (as long as attempts are logged).

**Comment by jsravn** [comment]
Okay, I can make that change 👍

**Comment by dashpole** [comment]
This will have to wait for #1868 before tests will pass.

**Comment by jsravn** [comment]
Okay, have a look.

File: manager/manager.go [hunk 1]
```diff
@@ -58,6 +59,7 @@ var logCadvisorUsage = flag.Bool("log_cadvisor_usage", false, "Whether to log th
 var eventStorageAgeLimit = flag.String("event_storage_age_limit", "default=24h", "Max length of time for which to store events (per type). Value is a comma separated list of key values, where the keys are event types (e.g.: creation, oom) or \"default\" and the value is a duration. Default is applied to all non-specified event types")
 var eventStorageEventLimit = flag.String("event_storage_event_limit", "default=100000", "Max number of events to store (per type). Value is a comma separated list of key values, where the keys are event types (e.g.: creation, oom) or \"default\" and the value is an integer. Default is applied to all non-specified event types")
 var applicationMetricsCountLimit = flag.Int("application_metrics_count_limit", 100, "Max number of application metrics to store (per container)")
+var dockerClientTimeout = flag.Duration("docker_client_timeout", 10*time.Second, "Timeout to the docker daemon")
```
Comment (dashpole) [inline manager/manager.go]: I would rather avoid adding a flag right now.  I think the retry mechanism is enough to fix the issue, and I don't think there is a meaningful decision for users to make here.

File: manager/manager.go [hunk 2]
```diff
@@ -234,6 +235,31 @@ func New(memoryCache *memory.InMemoryCache, sysfs sysfs.SysFs, maxHousekeepingIn
...
+	for {
+		ctx, _ := context.WithTimeout(context.Background(), startupTimeout)
+		dockerStatus, err := docker.StatusWithContext(ctx)
+		if err != nil {
+			return dockerStatus
+		}
+
+		switch err {
+		case context.DeadlineExceeded:
+			glog.Warningf("Timeout trying to communicate with docker, will retry")
```
Comment (dashpole) [inline manager/manager.go]: nit: add "during initialization" so we know this is delaying startup

**Comment by jsravn** [comment]
@dashpole Addressed your comments.

**Comment by dashpole** [comment]
you will have to rebase before tests will work
/ok-to-test

**Comment by jsravn** [comment]
@dashpole Okay, rebased.

**Pull request merged**